File: 2023/15/15.rb
#!/usr/bin/env ruby

sequence = $stdin.read.strip.split(',')

def hash_code(seq)
  val = 0
  seq.each_char.map(&:ord).each do |ascii_code|
    val += ascii_code
    val = (val * 17) % 256
  end
  val
end

# Part 1
hsh = sequence.sum do |seq|
  hash_code(seq)
end
puts hsh

# Part 2
boxes = Hash.new { |h, k| h[k] = {} }
sequence.each do |seq|
  if seq.include? '-'
    label = seq.split('-').first
    box = hash_code(label)
    boxes[box].delete(label)
  else # includes '='
    label, n = seq.split('=')
    box = hash_code(label)
    boxes[box][label] = n.to_i
  end
end

sum = 0
boxes.each do |box, lenses|
  lenses.each_with_index do |(label, focal), slot|
    sum += (box + 1) * focal * (slot + 1)
  end
end
puts sum

File: 2023/15/15.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func ParseInput() []string {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	return strings.Split(scanner.Text(), ",")
}

func hashCode(str string) int {
	hsh := 0
	for _, c := range str {
		hsh += int(c)
		hsh = (hsh * 17) % 256
	}
	return hsh
}

func part1(sequences []string) int {
	sum := 0
	for _, seq := range sequences {
		sum += hashCode(seq)
	}
	return sum
}

type Lens struct {
	label string
	focal int
}

func buildBoxes(sequences []string) map[int][]Lens {
	boxes := map[int][]Lens{}
	for _, seq := range sequences {
		if strings.Contains(seq, "-") {
			label := strings.Split(seq, "-")[0]
			box := hashCode(label)
			i := slices.IndexFunc(boxes[box], func(lens Lens) bool {
				return lens.label == label
			})
			if i >= 0 {
				boxes[box] = slices.Delete(boxes[box], i, i+1)
			}
		} else {
			parts := strings.Split(seq, "=")
			label, num := parts[0], parts[1]
			box := hashCode(label)
			focal, _ := strconv.Atoi(num)
			i := slices.IndexFunc(boxes[box], func(lens Lens) bool {
				return lens.label == label
			})
			if i >= 0 {
				boxes[box][i].focal = focal
			} else {
				boxes[box] = append(boxes[box], Lens{label, focal})
			}
		}
	}
	return boxes
}

func part2(sequences []string) int {
	boxes := buildBoxes(sequences)
	sum := 0
	for box, lenses := range boxes {
		for slot, lens := range lenses {
			sum += (box + 1) * lens.focal * (slot + 1)
		}
	}
	return sum
}

func main() {
	sequences := ParseInput()
	fmt.Println(part1(sequences))
	fmt.Println(part2(sequences))
}
